pkg/multicluster/watcher.go

```
@@ -0,0 +1,140 @@
+// It starts and runs ClusterHandlers for the localManager as if it were discovered by the watcher.
+func RunClusterWatcher(ctx context.Context, localManager manager.Manager, handlers ...ClusterHandler) error {
+	watcher := &clusterWatcher{
+		ctx:      ctx,
+		handlers: handlers,
+		cancels:  newCancelSet(),
+	}
+
+	err := watcher.startManager(LocalCluster, localManager)
+	if err != nil {
+		return eris.Wrap(err, "failed to register local kube config with multicluster watcher")
+	}
+
+	loop := controller.NewSecretReconcileLoop("cluster watcher", localManager)
+	return loop.RunSecretReconciler(ctx, watcher, kubeconfig.SecretPredicate{})
+}
+
+func (c *clusterWatcher) ReconcileSecret(obj *v1.Secret) (reconcile.Result, error) {
+	clusterName, cfg, err := kubeconfig.SecretToConfig(obj)
+	if err != nil {
+		return reconcile.Result{}, eris.Wrap(err, "failed to extract kubeconfig from secret")
+	}
+
+	if _, err := c.cancels.get(clusterName); err == nil {
+		return reconcile.Result{}, eris.Errorf("cluster %v already initialized", clusterName)
+	}
+
+	mgr, err := manager.New(cfg.RestConfig, manager.Options{
+		// TODO these should be configurable, disable for now
+		MetricsBindAddress:     "0",
+		HealthProbeBindAddress: "0",
+	})
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	return reconcile.Result{}, c.startManager(clusterName, mgr)
+}
+
+func (c *clusterWatcher) startManager(clusterName string, mgr manager.Manager) error {
+	ctx, cancel := context.WithCancel(
+		contextutils.WithLoggerValues(context.WithValue(c.ctx, "cluster", clusterName), zap.String("cluster", clusterName)))
+	go func() {
+		err := mgr.Start(ctx.Done())
+		if err != nil {
+			cancel()
+			contextutils.LoggerFrom(ctx).DPanicw("manager start failed for cluster %v", clusterName)
+		}
+	}()
+
+	c.cancels.set(clusterName, cancel)
+
+	errs := &multierror.Error{}
+	for _, handler := range c.handlers {
+		err := handler.HandleAddCluster(ctx, clusterName, mgr)
```

does this function need to return an error?